shim/src/lib.rs

```diff
@@ -0,0 +1,213 @@
+
+        #[link_name = "libtheseus::register_dtor"]
+        pub(crate) fn register_dtor(t: *mut u8, dtor: unsafe extern "C" fn(*mut u8));
+
+        #[link_name = "libtheseus::stdin"]
+        pub(crate) fn stdin() -> FfiResult<FatPointer, Error>;
+
+        #[link_name = "libtheseus::stdout"]
+        pub(crate) fn stdout() -> FfiResult<FatPointer, Error>;
+
+        #[link_name = "libtheseus::stderr"]
+        pub(crate) fn stderr() -> FfiResult<FatPointer, Error>;
+
+        #[link_name = "libtheseus::read"]
+        pub(crate) fn read(reader: FatPointer, buf: FfiSliceMut<'_, u8>)
+            -> FfiResult<usize, Error>;
+
+        #[link_name = "libtheseus::write"]
+        pub(crate) fn write(writer: FatPointer, buf: FfiSlice<'_, u8>) -> FfiResult<usize, Error>;
+
+        #[link_name = "libtheseus::flush"]
+        pub(crate) fn flush(writer: FatPointer) -> FfiResult<(), Error>;
+
+        #[link_name = "libtheseus::drop_reader"]
+        pub(crate) fn drop_reader(reader: FatPointer);
+
+        #[link_name = "libtheseus::drop_writer"]
+        pub(crate) fn drop_writer(writer: FatPointer);
+    }
+
+    const _: theseus_ffi::next_u64 = next_u64;
+    const _: theseus_ffi::getcwd = getcwd;
+    const _: theseus_ffi::chdir = chdir;
+    const _: theseus_ffi::getenv = getenv;
+    const _: theseus_ffi::setenv = setenv;
+    const _: theseus_ffi::unsetenv = unsetenv;
+    const _: theseus_ffi::exit = exit;
+    const _: theseus_ffi::getpid = getpid;
+    const _: theseus_ffi::register_dtor = register_dtor;
+    const _: theseus_ffi::stdin = stdin;
+    const _: theseus_ffi::stdout = stdout;
+    const _: theseus_ffi::stderr = stderr;
+    const _: theseus_ffi::read = read;
+    const _: theseus_ffi::write = write;
+    const _: theseus_ffi::flush = flush;
+    const _: theseus_ffi::drop_reader = drop_reader;
+    const _: theseus_ffi::drop_writer = drop_writer;
+}
+
+#[inline]
+pub fn next_u64() -> u64 {
+    unsafe { c::next_u64() }
+}
+
+#[inline]
+pub fn getcwd() -> String {
+    unsafe { c::getcwd() }.into()
+}
+
+#[inline]
+pub fn chdir(path: &str) -> Result<()> {
+    let path = path.into();
+    unsafe { c::chdir(path) }.into()
+}
+
+#[inline]
+pub fn getenv(key: &str) -> Option<String> {
+    let key = key.into();
+    Into::<Option<FfiString>>::into(unsafe { c::getenv(key) }).map(|s| s.into())
+}
+
+#[inline]
+pub fn setenv(key: &str, value: &str) -> Result<()> {
+    let key = key.into();
+    let value = value.into();
+    unsafe { c::setenv(key, value) }.into()
+}
+
+#[inline]
+pub fn unsetenv(key: &str) -> Result<()> {
+    let key = key.into();
+    unsafe { c::unsetenv(key) }.into()
+}
+
+#[inline]
+pub fn exit(code: i32) -> ! {
+    unsafe { c::exit(code) }
+}
+
+#[inline]
+pub fn getpid() -> u32 {
+    unsafe { c::getpid() }
+}
+
+#[inline]
+pub unsafe fn register_dtor(t: *mut u8, dtor: unsafe extern "C" fn(*mut u8)) {
+    c::register_dtor(t, dtor)
+}
+
+// TODO: impl Send + Sync for Reader and Writer?
```

Comment:
Bump. It should be fine because `FatPointer` is created from an `Arc`.

Reply:
is it necessary? as in, where is it needed for `Reader`/`Writer` to be `Send`/`Sync`?

If it is required, then imo it should be an impl on `FatPointer` rather than any other types that use it. If that's sound, then perhaps we should change `FatPointer` to indicate that it's a shared non-raw pointer that always comes from an `Arc`, e.g., `FatSharedPtr`

Reply:
No, I don't think it's necessary.